include/xsimd/config/xsimd_arch.hpp

```diff
@@ -25,147 +25,263 @@ namespace xsimd
+            struct contains<T> : std::false_type
+            {
+            };
 
-#if XSIMD_X86_AMD_INSTR_SET == XSIMD_X86_AMD_SSE4A_VERSION
-        using x86 = sse4a;
-#elif XSIMD_X86_AMD_INSTR_SET == XSIMD_X86_AMD_FMA4_VERSION
-        using x86 = fma4;
-#elif XSIMD_X86_AMD_INSTR_SET == XSIMD_X86_AMD_XOP_VERSION
-        using x86 = xop;
-#endif
+            template<class T, class Ty, class... Tys>
+            struct contains<T, Ty, Tys...> : std::conditional<std::is_same<Ty, T>::value, std::true_type, contains<T, Tys...>>::type
+            {
+            };
 
-/***********************
- * ARM instruction set *
- ***********************/
+            template<class... Archs>
+            struct issorted;
 
-#if XSIMD_ARM_INSTR_SET == XSIMD_ARM7_NEON_VERSION
-        using arm = neon;
-#elif XSIMD_ARM_INSTR_SET == XSIMD_ARM8_32_NEON_VERSION
-        using arm = neon;
-#elif XSIMD_ARM_INSTR_SET == XSIMD_ARM8_64_NEON_VERSION
-        using arm = neon64;
-#endif
+            template<>
+            struct issorted<> : std::true_type
+            {
+            };
+
+            template<class Arch>
+            struct issorted<Arch> : std::true_type
+            {
+            };
+
+            template<class A0, class A1, class... Archs>
+            struct issorted<A0, A1, Archs...> : std::conditional<(A0::version <= A1::version), issorted<Archs...>, std::false_type>::type
+            {
+            };
 
-#if XSIMD_X86_INSTR_SET != XSIMD_VERSION_NUMBER_NOT_AVAILABLE
-        using default_ = x86;
-#elif XSIMD_ARM_INSTR_SET != XSIMD_VERSION_NUMBER_NOT_AVAILABLE
-        using default_ = arm;
-#else
-        using default_ = scalar;
+        }
+
+
+        // An archlist is a list of architectures, sorted by version number.
+        template<class... Archs>
+        struct archlist {
+#ifndef NDEBUG
+          static_assert(detail::issorted<Archs...>::value, "architecture list must be sorted by version");
 #endif
 
+          template<class Arch>
+          using add = archlist<Archs..., Arch>;
+
+          template<class... OtherArchs>
+          using extend = archlist<Archs..., OtherArchs...>;
+
+          template<class Arch>
+          static constexpr bool contains() {
+            return detail::contains<Arch, Archs...>::value;
+          }
+        };
+
+
+        namespace detail
+        {
+            // Pick the best architecture in archlist L, which is the last
+            // because architectures are sorted by version.
+            template<class L>
+            struct best;
+
+            template<>
+            struct best<archlist<>>
+            {
+                using type = unavailable;
+            };
+
+            template<class Arch>
+            struct best<archlist<Arch>>
+            {
+                using type = Arch;
+            };
+
+            template<class Arch0, class Arch1, class... Archs>
+            struct best<archlist<Arch0, Arch1, Archs...>> : best<archlist<Arch1, Archs...>>
```

Comment:
Maybe it could be best to sort the archs in decreasing order. That would simplify the implementation of `best` (it can return the first one instead of having recursive instantiation of template structs).